processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndexStorageAdapter.java

```diff
@@ -142,24 +141,8 @@ public Comparable getMaxValue(String column)
   @Override
   public ColumnCapabilities getColumnCapabilities(String column)
   {
-    // Different from index.getCapabilities because, in a way, IncrementalIndex's string-typed dimensions
-    // are always potentially multi-valued at query time. (Missing / null values for a row can potentially be
-    // represented by an empty array; see StringDimensionIndexer.IndexerDimensionSelector's getRow method.)
-    //
-    // We don't want to represent this as having-multiple-values in index.getCapabilities, because that's used
-    // at index-persisting time to determine if we need a multi-value column or not. However, that means we
-    // need to tweak the capabilities here in the StorageAdapter (a query-time construct), so at query time
-    // they appear multi-valued.
-
-    final ColumnCapabilities capabilitiesFromIndex = index.getCapabilities(column);
-    final IncrementalIndex.DimensionDesc dimensionDesc = index.getDimension(column);
-    if (dimensionDesc != null && dimensionDesc.getCapabilities().getType() == ValueType.STRING) {
-      final ColumnCapabilitiesImpl retVal = ColumnCapabilitiesImpl.copyOf(capabilitiesFromIndex);
-      retVal.setHasMultipleValues(true);
-      return retVal;
-    } else {
-      return capabilitiesFromIndex;
-    }
+    // snapshot the current state
+    return ColumnCapabilitiesImpl.complete(index.getCapabilities(column), false);
```

Comment:
`StringDimensionIndexer.IndexerDimensionSelector` will not have `hasMultipleValues` set until it processes a multi-value row, is there a potential race between the time we look at capabilities that it doesn't yet have a multi-value row, but at execution time the row has appeared and so it will be set and will return an empty array instead of an array with a null?

Reply:
Yes, that's what I was thinking: HMV could be set at any time, including between cap-fetch and selector-creation.

By the way, it looks like StringDimensionIndexer's handling of `hasMultipleValues` is not thread-safe (it's updated by the indexing thread and read by querying threads without locking).

Reply:
Ok, given the race condition I guess I have to revert this behavior so queries use the fake capabilities and to fix the bug for segment metadata based schema discovery to plumb special the actual capabilities to segment metadata queries. Maybe longer term it would be better if we could snapshot at the time we query and make a cursor to only read the rows that were available at the time the capabilities were fetched so that we can have more performant queries on incremental indexes at the cost of missing a few rows.